### PR TITLE
INTMDB-475 Cloud backup snapshot export bucket

### PR DESCRIPTION
**Pull request by Adelmar92**
## Description

Added new Resource Cloud Backup export bucket:

The exportBuckets resource allows you to grant Atlas access to the specified bucket for exporting backup snapshots.

How to test:
In order to test this resource son Prerequisites are required, check [prerequisites](https://www.mongodb.com/docs/atlas/backup/cloud-backup/export/#prerequisites)

1. M10 or higher Atlas cluster with Cloud Backup [enabled.](https://www.mongodb.com/docs/atlas/backup/cloud-backup/overview/#std-label-backup-cloud-provider)

2. AWS [IAM role](https://docs.aws.amazon.com/IAM/latest/UserGuide/id_roles_use.html) with STS:AssumeRole that grants Atlas access to your AWS resources. To learn more about configuring AWS access for Atlas, see [Set Up Unified AWS Access.](https://www.mongodb.com/docs/atlas/security/set-up-unified-aws-access/#std-label-set-up-unified-aws-access)

3. AWS [IAM role policy](https://docs.aws.amazon.com/IAM/latest/UserGuide/access_policies.html) that grants Atlas write access or the S3:PutObject and S3:GetBucketLocation permissions to your AWS resources. To learn more about configuring write access to AWS resources, see [Set Up Unified AWS Access.](https://www.mongodb.com/docs/atlas/security/set-up-unified-aws-access/#std-label-set-up-unified-aws-access)

Two main resource are required:

1. an Atlas and a Aws role related by a trusted relationship, and permissions to access an s3 bucket
2. an s3 bucket (the bucket name is required)

Screenshot 2022-11-25 at 8 25 09 AM

Can be manually executed by using this json inputs and the command:

`cfn invoke resource {CREATE | READ | UPDATE (NOT SUPPORTED) | DELETE | LIST} file-name.json`
    
CREATE
```json
{
  "desiredResourceState": {
    {
  "GroupId": {GROUP-ID},
  "IamRoleID": {ROLE-ID-CREATED},
  "BucketName": {AWS-S3-BUCKET-NAME}
  "ApiKeys": {
    "PublicKey": {PUBLIC-KEY},
    "PrivateKey": {PRIVATE-KEY}
  }
}
  },
  "previousResourceState": {
  }
}
```
    
READ
```json
{
  "desiredResourceState": {
    {
  "Id": {EXPORT-BUCKET-ID},
  "GroupId": {GROUP-ID},
  "IamRoleID": {ROLE-ID-CREATED},
  "BucketName": {AWS-S3-BUCKET-NAME}
  "ApiKeys": {
    "PublicKey": {PUBLIC-KEY},
    "PrivateKey": {PRIVATE-KEY}
  }
}
  },
  "previousResourceState": {
  }
}
```

DELETE
```json
{
  "desiredResourceState": {
    {
  "Id": {EXPORT-BUCKET-ID},
  "GroupId": {GROUP-ID},
  "IamRoleID": {ROLE-ID-CREATED},
  "BucketName": {AWS-S3-BUCKET-NAME}
  "ApiKeys": {
    "PublicKey": {PUBLIC-KEY},
    "PrivateKey": {PRIVATE-KEY}
  }
}
  },
  "previousResourceState": {
  }
}
```

LIST
```json
{
  "desiredResourceState": {
    {
  "GroupId": {GROUP-ID},
  "IamRoleID": {ROLE-ID-CREATED},
  "BucketName": {AWS-S3-BUCKET-NAME}
  "ApiKeys": {
    "PublicKey": {PUBLIC-KEY},
    "PrivateKey": {PRIVATE-KEY}
  }
}
  },
  "previousResourceState": {
  }
}
```

Link to any related issue(s): 

## Type of change:

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

## Required Checklist:

- [x] I have signed the [MongoDB CLA](https://www.mongodb.com/legal/contributor-agreement)
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have added any necessary documentation (if appropriate)
- [x] I have run `make fmt` and formatted my code

### CODE REVIEW

File: cfn-resources/cloud-backup-snapshot-export-bucket/Makefile [hunk 1]
```diff
@@ -0,0 +1,23 @@
+.PHONY: build test clean
+tags=logging callback metrics scheduler
+cgo=0
+goos=linux
+goarch=amd64
+#export TAGS
+#export CGO_ENABLED
+#export GOARCH
```
Comment (andreaangiolillo) [inline cfn-resources/cloud-backup-snapshot-export-bucket/Makefile:8]:
```suggestion
```

File: cfn-resources/cloud-backup-snapshot-export-bucket/Makefile [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+# Set next value at compile time for fallback log-level
+ldXflags=github.com/mongodb/mongodbatlas-cloudformation-resources/util.defaultLogLevel=warning
+ldXflagsD=github.com/mongodb/mongodbatlas-cloudformation-resources/util.defaultLogLevel=debug
+build:
+	#make -f makebuild  # this runs build steps required by the cfn cli
```
Comment (andreaangiolillo) [inline cfn-resources/cloud-backup-snapshot-export-bucket/Makefile:13]:
```suggestion
```

File: cfn-resources/cloud-backup-snapshot-export-bucket/Makefile [hunk 3]
```diff
@@ -0,0 +1,23 @@
...
+	cfn generate
+	env GOOS=$(goos) CGO_ENABLED=$(cgo) GOARCH=$(goarch) go build -ldflags="-s -w -X '$(ldXflags)'" -tags="$(tags)" -o bin/handler cmd/main.go
+
+debug:
+	#make -f makebuild  # this runs build steps required by the cfn cli
```
Comment (andreaangiolillo) [inline cfn-resources/cloud-backup-snapshot-export-bucket/Makefile:18]:
```suggestion
```

File: cfn-resources/cloud-backup-snapshot-export-bucket/cmd/resource/resource.go [hunk 4]
```diff
@@ -0,0 +1,192 @@
...
+	progressevents "github.com/mongodb/mongodbatlas-cloudformation-resources/util/progressevent"
+
+	"github.com/aws-cloudformation/cloudformation-cli-go-plugin/cfn/handler"
+	"github.com/aws/aws-sdk-go/service/cloudformation"
+	local_constants "github.com/mongodb/mongodbatlas-cloudformation-resources/cloud-backup-snapshot-export-bucket/cmd/constants"
+	"github.com/mongodb/mongodbatlas-cloudformation-resources/util"
+	"github.com/mongodb/mongodbatlas-cloudformation-resources/util/constants"
+	log "github.com/mongodb/mongodbatlas-cloudformation-resources/util/logger"
+	"github.com/mongodb/mongodbatlas-cloudformation-resources/util/validator"
+	"go.mongodb.org/atlas/mongodbatlas"
```
Comment (andreaangiolillo) [inline cfn-resources/cloud-backup-snapshot-export-bucket/cmd/resource/resource.go:16]:
```suggestion
	progressevents "github.com/mongodb/mongodbatlas-cloudformation-resources/util/progressevent"
	"github.com/aws-cloudformation/cloudformation-cli-go-plugin/cfn/handler"
	"github.com/aws/aws-sdk-go/service/cloudformation"
	local_constants "github.com/mongodb/mongodbatlas-cloudformation-resources/cloud-backup-snapshot-export-bucket/cmd/constants"
	"github.com/mongodb/mongodbatlas-cloudformation-resources/util"
	"github.com/mongodb/mongodbatlas-cloudformation-resources/util/constants"
	log "github.com/mongodb/mongodbatlas-cloudformation-resources/util/logger"
	"github.com/mongodb/mongodbatlas-cloudformation-resources/util/validator"
	"go.mongodb.org/atlas/mongodbatlas"
```

File: cfn-resources/cloud-backup-snapshot-export-bucket/cmd/resource/resource.go [hunk 5]
```diff
@@ -0,0 +1,192 @@
...
+func Delete(req handler.Request, prevModel *Model, currentModel *Model) (handler.ProgressEvent, error) {
+	setup()
+
+	_, _ = log.Warnf("Create cluster model : %+v", currentModel)
+
+	// Validation
+	modelValidation := validateModel(DeleteRequiredFields, currentModel)
+	if modelValidation != nil {
+		return *modelValidation, nil
+	}
```
Comment (andreaangiolillo) [inline cfn-resources/cloud-backup-snapshot-export-bucket/cmd/resource/resource.go:128]:
```suggestion
	if modelValidation := validateModel(DeleteRequiredFields, currentModel); modelValidation != nil {
		return *modelValidation, nil
	}
```

**Review by andreaangiolillo — APPROVED** [review]
LGTM

File: cfn-resources/cloud-backup-snapshot-export-bucket/cmd/resource/resource.go (hunk 5 above)
Comment (Adelmar92) [reply cfn-resources/cloud-backup-snapshot-export-bucket/cmd/resource/resource.go:127]: awesome! thanks for the refactor

File: examples/cloud-backup-snapshot-export-bucket/privateEndpoint.json [hunk 6]
```diff
@@ -0,0 +1,41 @@
+{
+  "AWSTemplateFormatVersion" : "2010-09-09",
```
Comment (Zuhairahmed) [inline examples/cloud-backup-snapshot-export-bucket/privateEndpoint.json:2]: this filename says privateEndpoint.json should we update this for cloud-backup-snapshot-export-bucket.json example instead?

File: cfn-resources/cloud-backup-snapshot-export-bucket/README.md [hunk 7]
```diff
@@ -0,0 +1,60 @@
+# MongoDB::Atlas::CloudBackupSnapshotExportBucket
+
+The exportBuckets resource allows you to grant Atlas access to the specified bucket for exporting backup snapshots.
+In order tu create the resource some prerequisites are required, check [Export Cloud Backup Snapshot Docs](https://www.mongodb.com/docs/atlas/backup/cloud-backup/export/) for more information
```
Comment (Zuhairahmed) [inline cfn-resources/cloud-backup-snapshot-export-bucket/README.md:4]:
```suggestion
In order to create the resource some prerequisites are required, check [Export Cloud Backup Snapshot Docs](https://www.mongodb.com/docs/atlas/backup/cloud-backup/export/) for more information
```

File: cfn-resources/cloud-backup-snapshot-export-bucket/README.md [hunk 8]
```diff
@@ -0,0 +1,60 @@
...
+
+Both CREATE & DELETE tests must pass.
+
+## Installation
+TAGS=logging make
+cfn submit --verbose --set-default
+
+## Usage
+
+The [/quickstart-mongodb-atlas/scripts/launch-x-quickstart.sh](launch-x-quickstart.sh) script
```
Comment (Zuhairahmed) [inline cfn-resources/cloud-backup-snapshot-export-bucket/README.md:37]: link doesn't work, can you fix it please?

File: cfn-resources/cloud-backup-snapshot-export-bucket/README.md [hunk 9]
```diff
@@ -0,0 +1,60 @@
...
+
+Please consult the [Resource Docs](docs/README.md)
+
+## Local Testing
+
+The local tests are integrated with the AWS `sam local` and `cfn invoke` tooling features:
+
+```
+sam local start-lambda --skip-pull-image
+```
```
Comment (Zuhairahmed) [inline cfn-resources/cloud-backup-snapshot-export-bucket/README.md:16]: Missing section + link to Examples folder

**Review by Zuhairahmed — COMMENTED** [review]
Thanks @Adelmar92! few comments below:

**Review by Zuhairahmed — APPROVED** [review]
LGTM! Thanks @Adelmar92 for addressing all the feedback

**Review by ParthasarathyVaradhan — APPROVED** [review]
LGTM.

**Pull request merged**